### PR TITLE
JC-41

### PR DESCRIPTION
**Pull request by saffchen**

### CODE REVIEW

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+public class QuestionRepositoryCriteria {
+
+    @PersistenceContext
+    private final EntityManager entityManager;
+
+    public QuestionRepositoryCriteria(EntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    public Optional<List<QuestionEntity>> findByFilter(Pageable pageable, Difficulty difficulty, List<Tags> tags) {
```
Comment (pelmegov) [inline src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:28]: Do we need Pageable here? How we can use it?

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            Predicate tagsPredicate = tagsJoin.in(tags);
+            if (criteriaQuery.getRestriction() != null) {
+                criteriaQuery.where(criteriaBuilder.and(criteriaQuery.getRestriction(), tagsPredicate));
+            } else {
+                criteriaQuery.where(tagsPredicate);
+            }
+        }
+
+        criteriaQuery.orderBy(criteriaBuilder.asc(
+                criteriaBuilder.function("random", Integer.class)
```
Comment (pelmegov) [inline src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:51]: Minor: what is a random?

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        ));
+
+        List<QuestionEntity> resultList = entityManager.createQuery(criteriaQuery)
+                .setFirstResult((int) pageable.getOffset())
+                .setMaxResults(pageable.getPageSize())
+                .getResultList();
+
+        return Optional.ofNullable(resultList);
```
Comment (pelmegov) [inline src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:59]: Response can be nullable?

**Review by pelmegov — CHANGES_REQUESTED** [review]

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java (hunk 1 above)
Comment (saffchen) [reply src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:28]: To transfer the number of tasks we need to receive

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java (hunk 2 above)
Comment (saffchen) [reply src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:51]: Random order

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java (hunk 3 above)
Comment (saffchen) [reply src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java:59]: No, the response must always contain data.
Fixed

File: src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java (hunk 1 above)
Comment (pelmegov) [reply src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java]: Pageable usually used when you need separate something by pages, like 3 page by 10 elements, do you need such functionality here?

Comment (saffchen) [reply src/main/java/org/jacp/repository/QuestionRepositoryCriteria.java]: No, you can do without pageable. Fixed

**Review by pelmegov — APPROVED** [review]

**Review by freesky15 — APPROVED** [review]

**Pull request merged**